File: dashboard/app/Support/Footfall.php
<?php

declare(strict_types=1);

namespace App\Support;

use DateTimeImmutable;

/**
 * Stima dell'affluenza attesa, giorno per giorno.
 *
 * Porting dei pesi di crawler/src/osservatorio/analysis/footfall.py, che resta
 * il riferimento: se si cambiano i pesi vanno cambiati in entrambi i punti.
 * La parita' fra le due implementazioni e' verificata da dashboard/tests.
 *
 * I pesi sono un'ipotesi di partenza, non una misura: vanno ricalibrati sui
 * dati veri di stall_events (vedi la vista v_resa_punti_vendita).
 */
final class Footfall
{
    public const PESO_TIPOLOGIA = [
        'ipermercato'    => 1.00,
        'superstore'     => 0.85,
        'discount'       => 0.70,
        'supermercato'   => 0.65,
        'cash_and_carry' => 0.40,
        'superette'      => 0.35,
    ];
    public const PESO_TIPOLOGIA_IGNOTA = 0.50;

    /** Indici 1 = lunedi ... 7 = domenica, come DateTime::format('N'). */
    public const PESO_GIORNO = [1 => 0.55, 2 => 0.55, 3 => 0.60, 4 => 0.65, 5 => 0.85, 6 => 1.00, 7 => 0.60];

    public const PESO_LANCIO = 1.00;
    public const PESO_CENTRALE = 0.80;
    public const PESO_CHIUSURA = 0.90;
    public const PESO_SENZA_PROMO = 0.55;

    public const GIORNI_LANCIO = 3;
    public const GIORNI_CHIUSURA = 2;

    private const NOMI_GIORNI = [1 => 'lunedì', 2 => 'martedì', 3 => 'mercoledì',
        4 => 'giovedì', 5 => 'venerdì', 6 => 'sabato', 7 => 'domenica'];

    /** @return array{0: float, 1: string} peso e descrizione della fase */
    private static function fase(
        DateTimeImmutable $giorno,
        ?DateTimeImmutable $inizio,
        ?DateTimeImmutable $fine
    ): array {
        if ($inizio === null || $fine === null || $giorno < $inizio || $giorno > $fine) {
            return [self::PESO_SENZA_PROMO, 'nessuna promozione'];
        }
        if ((int) $inizio->diff($giorno)->days < self::GIORNI_LANCIO) {
            return [self::PESO_LANCIO, 'lancio campagna'];
        }
        if ((int) $giorno->diff($fine)->days < self::GIORNI_CHIUSURA) {
            return [self::PESO_CHIUSURA, 'chiusura campagna'];
        }
        return [self::PESO_CENTRALE, 'campagna in corso'];
    }

    public static function punteggio(
        DateTimeImmutable $giorno,
        ?string $tipologia = null,
        ?DateTimeImmutable $inizio = null,
        ?DateTimeImmutable $fine = null
    ): float {
        $pesoTipo = self::PESO_TIPOLOGIA[$tipologia] ?? self::PESO_TIPOLOGIA_IGNOTA;
        $pesoGiorno = self::PESO_GIORNO[(int) $giorno->format('N')];
        [$pesoFase] = self::fase($giorno, $inizio, $fine);

        return round(100 * $pesoTipo * $pesoGiorno * $pesoFase, 1);
    }

    /**
     * I giorni migliori dentro una finestra promozionale, dal piu' promettente.
     *
     * @return list<array{data: DateTimeImmutable, punteggio: float, fase: string, giorno: string}>
     */
    public static function giorniConsigliati(
        DateTimeImmutable $inizio,
        DateTimeImmutable $fine,
        ?string $tipologia = null,
        int $quanti = 3
    ): array {
        if ($fine < $inizio) {
            return [];
        }

        $candidati = [];
        for ($g = $inizio; $g <= $fine; $g = $g->modify('+1 day')) {
            [, $fase] = self::fase($g, $inizio, $fine);
            $candidati[] = [
                'data' => $g,
                'punteggio' => self::punteggio($g, $tipologia, $inizio, $fine),
                'fase' => $fase,
                'giorno' => self::NOMI_GIORNI[(int) $g->format('N')],
            ];
        }

        usort($candidati, static fn (array $a, array $b): int
            => [$b['punteggio'], $a['data']] <=> [$a['punteggio'], $b['data']]);

        return array_slice($candidati, 0, max(1, $quanti));
    }

    /** Solo per l'evidenza grafica: da che punteggio in su vale la pena andare. */
    public static function livello(float $punteggio): string
    {
        return match (true) {
            $punteggio >= 80 => 'alto',
            $punteggio >= 55 => 'medio',
            default => 'basso',
        };
    }
}

File: crawler/src/osservatorio/analysis/footfall.py
"""Stima dell'affluenza di un punto vendita, giorno per giorno.

Serve a una domanda operativa sola: in quale giorno conviene chiedere il
banchetto solidale, perche' il negozio sara' piu' frequentato e la raccolta
rendera' di piu'.

I pesi qui sotto sono un'IPOTESI di partenza, non una misura. Vanno ricalibrati
sui dati veri di `stall_events` (raccolto per banchetto, affluenza percepita)
appena ce ne sono a sufficienza: e' esattamente per questo che quella tabella
registra il raccolto e se c'era una promozione attiva.
"""

from __future__ import annotations

import dataclasses
import datetime as dt
import decimal

from osservatorio.models import StoreFormat

#: Quanto pesa la tipologia di punto vendita sul transito di persone.
PESO_TIPOLOGIA: dict[StoreFormat, float] = {
    StoreFormat.IPERMERCATO: 1.00,
    StoreFormat.SUPERSTORE: 0.85,
    StoreFormat.DISCOUNT: 0.70,
    StoreFormat.SUPERMERCATO: 0.65,
    StoreFormat.CASH_AND_CARRY: 0.40,
    StoreFormat.SUPERETTE: 0.35,
}
PESO_TIPOLOGIA_IGNOTA = 0.50

#: Giorno della settimana, 0 = lunedi'. Il sabato resta il giorno della spesa.
PESO_GIORNO: dict[int, float] = {
    0: 0.55, 1: 0.55, 2: 0.60, 3: 0.65, 4: 0.85, 5: 1.00, 6: 0.60,
}

#: Fase della campagna promozionale.
PESO_LANCIO = 1.00        # primi giorni: il volantino e' appena uscito
PESO_CENTRALE = 0.80
PESO_CHIUSURA = 0.90      # ultimi giorni: effetto ultima occasione
PESO_SENZA_PROMO = 0.55   # giorno senza campagna attiva

GIORNI_LANCIO = 3
GIORNI_CHIUSURA = 2


@dataclasses.dataclass(frozen=True, slots=True)
class GiornoConsigliato:
    """Un giorno candidato per il banchetto, con il perche'."""

    data: dt.date
    punteggio: float
    fase: str

    @property
    def giorno_settimana(self) -> str:
        nomi = ["lunedi", "martedi", "mercoledi", "giovedi", "venerdi", "sabato", "domenica"]
        return nomi[self.data.weekday()]


def _peso_fase(giorno: dt.date, inizio: dt.date | None, fine: dt.date | None) -> tuple[float, str]:
    if inizio is None or fine is None or not (inizio <= giorno <= fine):
        return PESO_SENZA_PROMO, "nessuna promozione"
    if (giorno - inizio).days < GIORNI_LANCIO:
        return PESO_LANCIO, "lancio campagna"
    if (fine - giorno).days < GIORNI_CHIUSURA:
        return PESO_CHIUSURA, "chiusura campagna"
    return PESO_CENTRALE, "campagna in corso"


def _arrotonda(valore: float) -> float:
    """Arrotonda a un decimale con la regola del mezzo per eccesso.

    round() di Python arrotonda al pari (55.25 -> 55.2), PHP per eccesso
    (55.3). Senza questa uniformita' il crawler e la dashboard mostrerebbero
    punteggi diversi per lo stesso giorno: la parita' e' verificata da
    crawler/tests/test_footfall_parity.py.
    """
    return float(
        decimal.Decimal(str(valore)).quantize(
            decimal.Decimal("0.1"), rounding=decimal.ROUND_HALF_UP
        )
    )


def punteggio_affluenza(
    giorno: dt.date,
    tipologia: StoreFormat | None = None,
    inizio_campagna: dt.date | None = None,
    fine_campagna: dt.date | None = None,
) -> float:
    """Punteggio 0-100 dell'affluenza attesa in un dato giorno.

    E' una stima comparativa: serve a ordinare i giorni e i punti vendita fra
    loro, non a prevedere quante persone entreranno davvero.
    """
    peso_tipo = PESO_TIPOLOGIA.get(tipologia, PESO_TIPOLOGIA_IGNOTA)
    peso_giorno = PESO_GIORNO[giorno.weekday()]
    peso_fase, _ = _peso_fase(giorno, inizio_campagna, fine_campagna)
    return _arrotonda(100 * peso_tipo * peso_giorno * peso_fase)


def giorni_consigliati(
    inizio_campagna: dt.date,
    fine_campagna: dt.date,
    tipologia: StoreFormat | None = None,
    quanti: int = 3,
) -> list[GiornoConsigliato]:
    """I giorni migliori per il banchetto dentro una finestra promozionale.

    Restituisce i candidati ordinati dal piu' promettente, cosi' la segretaria
    puo' proporre al punto vendita una data prima e un paio di alternative.
    """
    if fine_campagna < inizio_campagna:
        raise ValueError("la fine della campagna precede l'inizio")
    if quanti < 1:
        raise ValueError("serve almeno un giorno")

    candidati = []
    giorno = inizio_campagna
    while giorno <= fine_campagna:
        _, fase = _peso_fase(giorno, inizio_campagna, fine_campagna)
        candidati.append(
            GiornoConsigliato(
                data=giorno,
                punteggio=punteggio_affluenza(
                    giorno, tipologia, inizio_campagna, fine_campagna
                ),
                fase=fase,
            )
        )
        giorno += dt.timedelta(days=1)

    candidati.sort(key=lambda c: (-c.punteggio, c.data))
    return candidati[:quanti]
